File: 백준/Gold/17471. 게리맨더링/게리맨더링.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

public class Main {

	static Integer[] parents;
	static boolean[] isSelected;
	static int N;
	static int[] p;
	static int[][] arr;
	static int result = Integer.MAX_VALUE;

	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

		N = Integer.parseInt(br.readLine());
		parents = new Integer[N];
		isSelected = new boolean[N];

		p = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();

		arr = new int[N][];
		for (int i = 0; i < N; i++) {
			arr[i] = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
		}
		subset(0);
		System.out.println(result == Integer.MAX_VALUE ? -1 : result);
	}

	static void makeSet(int v) {
		for (int i = 0; i < v; i++) {
			parents[i] = i;
		}
	}

	static int findSet(int v) {
		if (v == parents[v])
			return v;
		return parents[v] = findSet(parents[v]);
	}

	static void union(int a, int b) {
		int aRoot = findSet(a);
		int bRoot = findSet(b);

		if (aRoot < bRoot)
			parents[bRoot] = aRoot;
		else
			parents[aRoot] = bRoot;
	}

	public static void subset(int cnt) {
		if (cnt == N) {
			makeSet(N);
			for (int i = 0; i < N; i++) {
				if (isSelected[i]) {
					for (int j = 1, size = arr[i][0]; j <= size; j++) {
						if (isSelected[arr[i][j] - 1])
							union(i, arr[i][j] - 1);
					}
				} else {
					for (int j = 1, size = arr[i][0]; j <= size; j++) {
						if (!isSelected[arr[i][j] - 1])
							union(i, arr[i][j] - 1);
					}
				}
			}
			check();
			return;
		}
		isSelected[cnt] = true;
		subset(cnt + 1);
		isSelected[cnt] = false;
		subset(cnt + 1);
	}

	private static void check() {
//		Set<Integer> set = new HashSet<>(Arrays.asList(parents));
		Set<Integer> set = new HashSet<>();
		for (int i = 0; i<N; i++) {
			set.add(findSet(i));
		}

		if (set.size() == 2) {
			int a = 0, b = 0;
			int flag = parents[0];
			for (int i = 0; i < N; i++) {
				if (parents[i] == flag)
					a += p[i];
				else
					b += p[i];
			}
			result = Math.min(result, Math.abs(a - b));
		}
	}
}

File: 백준/Gold/17471. 게리맨더링/게리맨더링.py
import sys

input = sys.stdin.readline


def subset(cnt):
    global isSelected
    if cnt == N:
        makeSet(N)
        for i in range(1, N + 1):
            if isSelected[i]:
                for j in range(len(arr[i])):
                    if isSelected[arr[i][j]]:
                        union(i, arr[i][j])
            else:
                for j in range(len(arr[i])):
                    if not isSelected[arr[i][j]]:
                        union(i, arr[i][j])
        check()
        return
    isSelected[cnt] = True
    subset(cnt + 1)
    isSelected[cnt] = False
    subset(cnt + 1)


def check():
    global result, parents
    s = set()
    for i in range(1, N + 1):
        s.add(findSet(i))

    if len(s) == 2:
        a = 0
        b = 0
        flag = parents[1]
        for i in range(1, N + 1):
            if parents[i] == flag:
                a += p[i - 1]
            else:
                b += p[i - 1]
        result = min(result, abs(a - b))


def makeSet(n):
    global parents
    for i in range(1, n + 1):
        parents[i] = i


def findSet(x):
    global parents
    if x == parents[x]:
        return x
    parents[x] = findSet(parents[x])
    return parents[x]


def union(a, b):
    global parents
    a = findSet(a)
    b = findSet(b)
    if a < b:
        parents[b] = a
    else:
        parents[a] = b


N = int(input())
p = list(map(int, input().split()))
isSelected = [False] * (N + 1)
parents = [0] * (N + 1)

arr = [[]]
for _ in range(N):
    arr.append(list(map(int, input().split()))[1:])

result = 1e9
subset(0)
print(result if result != 1e9 else -1)